### PR TITLE
KAFKA-9074: Correct Connect’s `Values.parseString` to properly parse a time and timestamp literal

### PR DESCRIPTION
**Pull request by rhauch**
Time and timestamp literal strings contain a `:` character, but the internal parser used in the `Values.parseString(String)` method tokenizes on the colon character to tokenize and parse map entries. The colon could be escaped, but then the backslash character used to escape the colon is not removed and the parser fails to match the literal as a time or timestamp value.

This fix corrects the parsing logic to properly parse timestamp and time literal strings whose colon characters are either escaped or unescaped. Additional unit tests were added to first verify the incorrect behavior and then to validate the correction.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (ncliang) [inline connect/api/src/main/java/org/apache/kafka/connect/data/Values.java]: Can these be default visibility instead of protected?

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (ncliang) [inline connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:1176]: Calling nextTokens with negative n would cause infinite loop.

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (ncliang) [inline connect/api/src/main/java/org/apache/kafka/connect/data/Values.java]: This interface seems strange to me. Any reason you want this separately instead of just doing this in nextTokens(int) above?

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (ncliang) [inline connect/api/src/main/java/org/apache/kafka/connect/data/Values.java]: Maybe make private?

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (ncliang) [inline connect/api/src/main/java/org/apache/kafka/connect/data/Values.java]: Maybe just call it `next(int)` to match existing `next()` method?

**Review by ncliang — COMMENTED** [review]
Looks good in general, with good test coverage! Just a few comments I have.

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (rhauch) [reply connect/api/src/main/java/org/apache/kafka/connect/data/Values.java]: ~Why is default visibility better?~

Okay, it is a bit unclear why things are `protected static`. Changed.

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (rhauch) [reply connect/api/src/main/java/org/apache/kafka/connect/data/Values.java:1176]: ~Added a check.~

Actually removed this method variant. See next comment.

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (rhauch) [reply connect/api/src/main/java/org/apache/kafka/connect/data/Values.java]: When we need to call this, we've already consumed the first token, so the idea was to be able to supply the starting position for the substring.

However, you have a valid point. I can remove this method altogether and in the parser method where this was called simply get the n-1 tokens and append to the current token. That makes the parser method a tiny bit more complicated, but I think it's more obvious what is going on.

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (rhauch) [reply connect/api/src/main/java/org/apache/kafka/connect/data/Values.java]: How about package-protected?

File: connect/api/src/main/java/org/apache/kafka/connect/data/Values.java
Comment (rhauch) [reply connect/api/src/main/java/org/apache/kafka/connect/data/Values.java]: Changed.

**Review by ncliang — APPROVED** [review]
LGTM 👍

**Comment by C0urante** [comment]
@rhauch the changes here look good but I think they're insufficient for nested values. I've written a couple of failing tests that fail on the code in this PR to demonstrate my point:

```java
    @Test
    public void shouldParseDateStringAsDateInArray() throws Exception {
        String dateStr = "2019-08-23";
        String arrayStr = "[" + dateStr + "]";
        SchemaAndValue result = Values.parseString(arrayStr);
        assertEquals(Type.ARRAY, result.schema().type());
        Schema elementSchema = result.schema().valueSchema();
        assertEquals(Type.INT32, elementSchema.type());
        assertEquals(Date.LOGICAL_NAME, elementSchema.name());
        java.util.Date expected = new SimpleDateFormat(Values.ISO_8601_DATE_FORMAT_PATTERN).parse(dateStr);
        assertEquals(Collections.singletonList(expected), result.value());
    }

    @Test
    public void shouldParseTimeStringAsDateInArray() throws Exception {
        String dateStr = "14:34:54.346Z";
        String arrayStr = "[" + dateStr + "]";
        SchemaAndValue result = Values.parseString(arrayStr);
        assertEquals(Type.ARRAY, result.schema().type());
        Schema elementSchema = result.schema().valueSchema();
        assertEquals(Type.INT32, elementSchema.type());
        assertEquals(Date.LOGICAL_NAME, elementSchema.name());
        java.util.Date expected = new SimpleDateFormat(Values.ISO_8601_TIME_FORMAT_PATTERN).parse(dateStr);
        assertEquals(Collections.singletonList(expected), result.value());
    }

    @Test
    public void shouldParseTimeStringAsDateInMap() throws Exception {
        String dateStr = "14:34:54.346Z";
        String arrayStr = "{3:" + dateStr + "}";
        SchemaAndValue result = Values.parseString(arrayStr);
        assertEquals(Type.MAP, result.schema().type());
        Schema keySchema = result.schema().keySchema();
        assertEquals(Type.INT8, keySchema.type());
        Schema valueSchema = result.schema().valueSchema();
        assertEquals(Type.INT32, valueSchema.type());
        assertEquals(Date.LOGICAL_NAME, valueSchema.name());
        java.util.Date expected = new SimpleDateFormat(Values.ISO_8601_TIME_FORMAT_PATTERN).parse(dateStr);
        assertEquals(Collections.singletonMap((byte) 3, expected), result.value());
    }
```

If these cases (e.g., use as elements in arrays/maps) should be handled for times and timestamp literals, can we address them in this PR as well?

**Comment by rhauch** [comment]
Thanks, @C0urante. I've added your suggested tests plus a few more.

**Review by C0urante — APPROVED** [review]
LGTM, thanks for the added coverage @rhauch

**Comment by rhauch** [comment]
Retest this, please.

**Comment by rhauch** [comment]
The previous build passed on JDK 8, and failures on other two were unrelated. But retesting anyway.

@hachikuji, would you mind reviewing this? #7284 is blocked by this. Thanks!

**Comment by rhauch** [comment]
The JDK 11 / Scala 2.13 had this unrelated failure:
```
Test Result (1 failure / +1)
  kafka.api.SslAdminClientIntegrationTest.testSynchronousAuthorizerAclUpdatesBlockRequestThreads
```

**Comment by alozano3** [comment]
What's the status of this PR? Is this ready to merge once the flaky tests pass? I am waiting for this in order to finish the PR [#7284](https://github.com/apache/kafka/pull/7284)
Regards.

**Comment by rhauch** [comment]
@alozano3, we missed AK 2.4 code freeze for this PR, and because it needs to be backported to older branches we're waiting to merge this to `master` until we can backport to the `2.4` branch.

**Comment by rhauch** [comment]
Retest this please

**Comment by rhauch** [comment]
Rebased on `trunk` to fix conflicts.

The original logic was introduced in AK 1.1.0, but since we typically backport only 2-3 releases my plan is to backport this as far back as the `2.2` branch.

**Review by hachikuji — APPROVED** [review]
LGTM

**Pull request merged**
